packages/commonwealth/shared/canvas/serverVerify.ts

```diff
@@ -0,0 +1,141 @@
+import { verify } from './verify';
+import { import_ } from '@brillout/import';
+
+function assert(condition: unknown, message?: string): asserts condition {
+  if (!condition) {
+    throw new Error(message ?? 'assertion failed');
+  }
+}
+
+function assertMatches(a, b, obj: string, field: string) {
+  assert(
+    a === b,
+    `Invalid signed ${obj} (${field}: ${JSON.stringify(a)}, ${JSON.stringify(
+      b
+    )})`
+  );
+}
+
+const verifyUnpack = async (canvas_action, canvas_session, address) => {
+  const { actionType, sessionType } = await import_('@canvas-js/core/codecs');
+  const action = JSON.parse(canvas_action);
+  const session = JSON.parse(canvas_session);
+  assert(actionType.is(action), 'Invalid signed action (typecheck)');
+  assert(sessionType.is(session), 'Invalid signed session (typecheck)');
+  const [verifiedAction, verifiedSession] = await Promise.all([
+    verify({ action, actionSignerAddress: action.session }),
+    verify({ session }),
+  ]);
+  assert(verifiedAction === true, 'Invalid signed action (signature)');
+  assert(verifiedSession === true, 'Invalid signed session (signature)');
+  assert(
+    action.session === session.payload.sessionAddress,
+    'Invalid action/session pair'
+  );
+  return { action, session };
+};
+
+export const verifyComment = async (
+  canvas_action,
+  canvas_session,
+  canvas_hash,
+  fields
+) => {
+  if (
+    canvas_action === undefined &&
+    canvas_session === undefined &&
+    canvas_hash === undefined
+  )
+    return;
+  const { thread_id, text, address, chain, parent_comment_id } = fields;
```

Comment:
Is there a reason why I don't already use the types interface from canvas-node/actions here to verify the fields?

Reply:
Not yet, there will have to be another merge here, but I'd rather make that change later as this touches a main logic flow and any further changes increase the risk of breaking something.

Reply:
Possible to link / lmk when that gets in